### PR TITLE
Enable indexalator for device code

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Enables indexalator to be instantiated from device code.
Also add gtests for the output indexalator.
This change helps enable for the offset-normalizing-iterator #14234

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/include/cudf/detail/normalizing_iterator.cuh [hunk 1]
```diff
@@ -202,11 +202,32 @@ struct base_normalator {
...
+    constexpr int operator()() const
+    {
+#ifndef __CUDA_ARCH__
+      CUDF_FAIL("only integral types are supported");
+#else
+      CUDF_UNREACHABLE("only integral types are supported");
+#endif
+    }
+    template <typename T, std::enable_if_t<cudf::is_index_type<T>()>* = nullptr>
+    constexpr int operator()() const noexcept
```
Comment (bdice) [inline cpp/include/cudf/detail/normalizing_iterator.cuh:217]: `sizeof` returns a `size_t`. Should we do the same here? Or insert a cast?
```suggestion
    constexpr std::size_t operator()() const noexcept
```

**Review by bdice — APPROVED** [review]
Just one comment, otherwise LGTM.

File: cpp/include/cudf/detail/normalizing_iterator.cuh [hunk 2]
```diff
@@ -202,11 +202,32 @@ struct base_normalator {
     return static_cast<Derived const&>(*this).p_ >= rhs.p_;
   }
 
+ private:
+  struct integer_sizeof_fn {
+    template <typename T, std::enable_if_t<not cudf::is_index_type<T>()>* = nullptr>
+    constexpr int operator()() const
```
Comment (PointKernel) [inline cpp/include/cudf/detail/normalizing_iterator.cuh:208]:
```suggestion
   CUDF_HOST_DEVICE constexpr int operator()() const
```
nit: explicit `CUDF_HOST_DEVICE` constexpr according to #7795

**Review by PointKernel — APPROVED** [review]
LGTM

**Pull request merged**